### PR TITLE
#544 Longitudinal profile extension

### PR DESCRIPTION
**Pull request by alexander-fedorenko**
This PR provides extension that migrate the "Profil en long" plugin to the MapStore
https://github.com/georchestra/mapstore2-georchestra/issues/544

Key features:

- User is able to draw the line on the map to build profile out of it.
- User is able to import JSON file (map json export, annotations export or plain geoJSON format with features will work) to use line feature from it to build profile. Extension will search for a "LineString" or "MultiLineString" feature in file. Import will fail if suitable feature is not found.
- User is able to select feature on the map by a simple click when desired layer is selected in TOC. Selection will fail if clicked point has no line feature.
- User will see line drawn on the map when dock panel is open
- User is able to download results as CSV using button underneath the chart.
- User is able to download chart as png file.
- User is able to maximize chart to make it more suitable for exporting as an image. Maximized chart covers whole map viewport, sidebar panel and sidebar.
- User is able to see stats of the generated profile by switch to the second tab of dock panel.
- User is able to change preferences by using third tab of dock panel or via separate preferences dialog available in menu.

Extension can be built using `npm run ext:build` just like any other extension. Right now locally built version is installed on dev https://georchestra.geo-solutions.it/mapstore/#/context/Longitudinal

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
-    "name": "MapStoreExtension",
+    "name": "longitutdinal_profile",
```
Comment (dsuren1) [inline package.json:2]: Can you modify it to something like `LongitudinalProfile` ?

Comment (alexander-fedorenko) [reply package.json:2]: Sure, but it complains then that it doesn't match regex. That's the only reason I didn't use camel case 😄

**Review by dsuren1 — CHANGES_REQUESTED** [review]
Hi @alexander-fedorenko 
Thanks. Most features looks good to me.

However I found few things. Kindly look into it

- [x] Translations are missing for the following in **en.US** (`longitudinal.infos`, `longitudinal.chart`, `longitudinal.preferences`)
- [x] I see other extension made for georchestra app has **fr-FR** and **it-IT** translations too. Could you please check if it's required and add them if needed
- [ ] When `Selecting a profile`, upon clicking on a layer with linestring or multilinestring there is some parsing error when reprojecting. Ex: `Linea_costa` layer. Kindly check if it happens only with this layer. I checked it with meansurement converted as layer and annotations (works fine with these)
2022-07-06 17_32_44-DevTools - localhost_8091__debug=true


- [x] Could you add some padding to these elements, vertical spacing maybe. 

image

File: package.json (hunk 1 above)
Comment (dsuren1) [reply package.json:2]: Some weird npm conventions :)   Never mind. Leave it as it is

**Comment by alexander-fedorenko** [comment]
@dsuren1 @tdipisa Selection tool should work fine now. I've updated extension version on dev to the most recent one.
I think I will need a bit of time to test all edge cases related to projection of selected data or the file being uploaded or combination of non-standard map projection with drawing/selection tools.
@tdipisa provided geoJson has some exotic projection used. Do you have projectionDefs with "EPSG:26713" available? I just want to test it and make changes to the checks of the files to be sure that it works as expected if custom projection is supported and when it's not.

**Comment by dsuren1** [comment]
@alexander-fedorenko 
Try with this please
`Proj4js.defs("EPSG:26713", "+proj=utm +zone=13 +datum=NAD27 +units=m +no_defs");`

**Comment by alexander-fedorenko** [comment]
@tdipisa @dsuren1 with the recent changes following is added:
- Spinner in map toolbar shown whenever there is a WPS request to get profile data and whenever user clicks on the map to select feature.
- Selected line is highlighted, there was an issue with data reprojection.
- Same with chart data, it was projection issues.
- Uploading of JSON with data in custom projection will work as far as this projection is defined in localConfig -> projectionDefs. If it's not - import dialog will throw a warning that projection is not supported.

I didn't have a chance to do an additional check how it works when custom projection of the map is used. I'll check it as quickly as it possible and update the PR after that.

Also project have github workflow now to run linter, tests, build and provide an artifact.

**Comment by tdipisa** [comment]
Thank you so much @alexander-fedorenko can we proceed with a new review or should we wait for further checks you still have to do?

**Comment by alexander-fedorenko** [comment]
> Thank you so much @alexander-fedorenko
> 
> > I didn't have a chance to do an additional check how it works when custom projection of the map is used. I'll check it as quickly as it possible and update the PR after that.
> 
> Can we proceed with a new review or should we wait for further checks you still have to do?

I just did a quick test on dev and locally with map being in EPSG:4326 & EPSG:26713. All looks fine, drawing & selection tools working just fine, as well as importing.

@dsuren1 It can be reviewed now.

**Review by dsuren1 — CHANGES_REQUESTED** [review]
Hi @alexander-fedorenko 
Thanks for the update.

After recent commits, I stumbled upon these issue. Kindly look into it.
1. There seems to be some issue with active menu state. Menu active vs action performed are not in sync. (Maybe follow exact steps as shown in gif to reproduce)
profile_en_long
2. When an error occurs, the dropzone retains error on subsequent openings. It should be cleared.
dnd retention
3. If I'm not wrong, earlier when extension menu is toggled i.e `Draw Profile`, other map interaction tools were disabled/toggled off.
Ex: Query tool if enabled, upon toggling `Draw Profile`, the query tool was disabled
But now, I have observed this behavior, 
 - Enable `Draw Profile`
 - Draw a line on map
 - Enable Query tool from map toolbar
 - Toggle off `Draw Profile`
 - Click on Map
 - Query panel (GFI) is displayed (This is correct as extension plugin is not active now)
 - Re-enable `Draw Profile`
 - Draw a line on map
 - The GFI panel is not closed to show Longitudinal Profile panel 

**TL;DR** Upon toggling on `Draw Profile` & `Selection to profile`, other panels taking over the the space of `Longitudinal Profile` panel (GFI, measure.. etc) should be closed or in other words the _other map interaction tools need to disabled_
2022-07-11 13_19_03-Template project to create MapStore extensions

**Comment by alexander-fedorenko** [comment]
Hi @dsuren1 
1. You're right, action that should deactivate listening to the click event was not triggered in case if Selection tool was toggled to any other tool (drawing, import). This one is addressed now.
2. Thanks, import component will be unmounted whenever tools is deactivated. Previously state was persisted due to enhancers applied to the component that is always mounted.
3. Yes, right, that's still the case. Any tool that uses drawing interaction (measurement, annotations, urbanisme and cadastrapp extensions etc.) is active exclusively, any previously active tool will to be toggled off when new one is activated. Identify tool results is exception and it need to be deactivated explicitly. With the changes, longitudinal drawing, selection tools will be toggled off whenever identify is activated in map toolbar. Vise versa, identify tool will be deactivated and dock panel with query results will be closed once longitudinal tool is activated.

**Review by dsuren1 — CHANGES_REQUESTED** [review]
Hi @alexander-fedorenko 
Thanks for the update. We are almost there, just a few minor ones

- When the panel is closed the data shown in charts should be cleared as well
**Ex**: Chart is loaded with `profile data`, now upon loading new data for the drawn line, the chart shows previous data and then after a few sec is updated with new data. Can this be addressed?
profile_en

File: js/extension/epics/setup.js [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+                    console.log("Error while obtaining data for longitudinal profile"); // eslint-disable-line no-console
+                    console.log(e); // eslint-disable-line no-console
+                    return Rx.Observable.empty();
+                })
+                .let(wrapStartStop(
+                    [loading(true), ...(!isDockOpen(state) ? [openDock()] : [])],
+                    loading(false),
+                    () => Rx.Observable.of(error({
+                        title: "notification.error",
+                        message: "error loading data for longitudinal profile",
```
[the remark this answers is not in the document]
Comment (alexander-fedorenko) [reply js/extension/epics/setup.js]: Thanks, added!

File: js/extension/epics/selection.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+                        console.log("Error while obtaining data for longitudinal profile"); // eslint-disable-line no-console
+                        console.log(e); // eslint-disable-line no-console
+                        return Rx.Observable.of(loading(false));
+                    })
+                    .let(wrapStartStop(
+                        [loading(true)],
+                        [loading(false)],
+                        () => Rx.Observable.of(error({
+                            title: "notification.error",
+                            message: "error loading data for longitudinal profile",
```
[the remark this answers is not in the document]
Comment (alexander-fedorenko) [reply js/extension/epics/selection.js]: Thanks, added!

**Review by dsuren1 — APPROVED** [review]
LGTM
